pkg/autodiscovery/listeners/cloudfoundry.go

```diff
@@ -0,0 +1,235 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2017-2020 Datadog, Inc.
+
+// +build clusterchecks
+
+package listeners
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/autodiscovery/integration"
+	"github.com/DataDog/datadog-agent/pkg/util/cloudfoundry"
+)
+
+const (
+	CF_SERVICE_CONTAINER_IP = "container-ip"
+)
+
+type CloudFoundryListener struct {
+	newService chan<- Service
+	delService chan<- Service
+	services   map[string]Service // maps ADIdentifiers to services
+	stop       chan bool
+	t          *time.Ticker
+	m          sync.RWMutex
+	bbsCache   *cloudfoundry.BBSCache
+}
+
+type CloudFoundryService struct {
+	adIdentifier   cloudfoundry.ADIdentifier
+	containerIPs   map[string]string
+	containerPorts []ContainerPort
+	creationTime   integration.CreationTime
+}
+
+// Make sure CloudFoundryService implements the Service interface
+var _ Service = &CloudFoundryService{}
+
+func init() {
+	Register("cloudfoundry-bbs", NewCloudFoundryListener)
+}
+
+// NewCloudFoundryListener creates a CloudFoundryListener
+func NewCloudFoundryListener() (ServiceListener, error) {
+	bbsCache, err := cloudfoundry.GetGlobalBBSCache()
+	if err != nil {
+		return nil, err
+	}
+	return &CloudFoundryListener{
+		services: map[string]Service{},
+		stop:     make(chan bool),
+		t:        time.NewTicker(10 * time.Second),
+		bbsCache: bbsCache,
+	}, nil
+}
+
+// Listen periodically refreshes services from global BBS API cache
+func (l *CloudFoundryListener) Listen(newSvc chan<- Service, delSvc chan<- Service) {
+	// setup the I/O channels
+	l.newService = newSvc
+	l.delService = delSvc
+
+	go func() {
+		l.refreshServices(true)
+		for {
+			select {
+			case <-l.stop:
+				return
+			case <-l.t.C:
+				l.refreshServices(false)
+			}
+		}
+	}()
+}
+
+func (l *CloudFoundryListener) refreshServices(firstRun bool) {
+	l.bbsCache.RLock()
+	defer l.bbsCache.RUnlock()
```

Comment:
let's make all public methods of `bbsCache` thread-safe to avoid dealing with thread safety in the clients.

Reply:
So AFAICS all of the public methods of the cache are thread safe (IOW they acquire the read lock before returning a result and a returned result will never be modified further). The issue is when a method, such as this one, does multiple calls to various public methods of the cache - then we acquire the read lock here as well, to ensure that the method sees a consistent state of all the data structures obtained from the cache. For example, we don't want to get Desired LRPs from one run of the cache and Actual LRPs from a different run.

Does that make sense? If so, I can definitely provide more comments in the code to make this more obvious.

Reply:
Why can't we make `cloudfoundry.BBSCache` provide the required information directly without having to call multiple methods by the clients? e.g make `getAllADIdentifiers` a public `BBSCache` method and handle thread safety there. Just to avoid handling thread safety in the api AND in the clients.

Reply:
Yeah, that's a good point. I think I can do that. Thanks for the suggestion!

Reply:
Done.